lib/llvmopencl/Workgroup.cc

```diff
@@ -195,6 +195,52 @@ class WorkgroupImpl {
   unsigned long DeviceMaxWItemSizes[3];
 };
 
+/// Convert address space casts through integer casts to proper address space
+/// cast instructions.
+///
+/// At least the HIP/CUDA frontend since LLVM 19 seems to produce AS casts
+/// between generic and global ASs through ptrtoint - inttoptr conversion chains
+/// when targeting SPIR-V. It confuses the alias analyzer and leads to lack of
+/// utilization of "restrict" pointer information. It should be fixed upstream,
+/// but while that happens, let's just peephole convert it.
+static bool cleanupAddressSpaceCasts(Function &F) {
+  bool Changed = false;
+  for (auto &BB : F) {
+    for (BasicBlock::iterator BI = BB.begin(), BE = BB.end(); BI != BE;) {
+
+      // The casts look like this:
+      // %55 = ptrtoint ptr addrspace(1) %7 to i64
+      // %56 = inttoptr i64 %55 to ptr addrspace(4)
+
+      PtrToIntInst *PtrToInt = dyn_cast_or_null<PtrToIntInst>(&*BI);
```

Comment:
```suggestion
      PtrToIntInst *PtrToInt = dyn_cast<PtrToIntInst>(&*BI);
```
Unless the cast may receive `nullptr` values which probably is not the case.